Doesn't compile on 6.14 kernel (fedora 42)

hello, after upgrading my laptop to fedora 42 with kernel 6.14 this driver no longer compiles
```bash
                                                                                
~   
❯ cd /tmp
                                                                                
/tmp   
❯ git clone https://github.com/asymingt/rts5139.git
Cloning into 'rts5139'...
remote: Enumerating objects: 99, done.
remote: Counting objects: 100% (18/18), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 99 (delta 6), reused 13 (delta 5), pack-reused 81 (from 1)
Receiving objects: 100% (99/99), 253.17 KiB | 422.00 KiB/s, done.
Resolving deltas: 100% (39/39), done.
                                                                                
/tmp   
❯ cd rts5139 
                                                                                
rts5139 git:master  
❯ make
make -C /lib/modules/6.14.2-300.fc42.x86_64/build M=/tmp/rts5139 modules
make[1]: Entering directory '/usr/src/kernels/6.14.2-300.fc42.x86_64'

make[2]: Entering directory '/tmp/rts5139'
  CC [M]  rts51x_transport.o
  CC [M]  rts51x_scsi.o
rts51x_scsi.c: In function ‘rts51x_scsi_show_command’:
rts51x_scsi.c:328:77: warning: suggest braces around empty body in an ‘if’ statement [-Wempty-body]
  328 |                 RTS51X_DEBUGP("Command %s (%d bytes)\n", what, srb->cmd_len);
      |                                                                             ^
rts51x_scsi.c: At top level:
rts51x_scsi.c:2002:5: warning: no previous prototype for ‘queuecommand_lck’ [-Wmissing-prototypes]
 2002 | int queuecommand_lck(struct scsi_cmnd *srb)
      |     ^~~~~~~~~~~~~~~~
rts51x_scsi.c:2125:10: error: ‘struct scsi_host_template’ has no member named ‘slave_alloc’
 2125 |         .slave_alloc = slave_alloc,
      |          ^~~~~~~~~~~
rts51x_scsi.c:2125:24: error: initialization of ‘short unsigned int’ from ‘int (*)(struct scsi_device *)’ makes integer from pointer without a cast [-Wint-conversion]
 2125 |         .slave_alloc = slave_alloc,
      |                        ^~~~~~~~~~~
rts51x_scsi.c:2125:24: note: (near initialization for ‘rts51x_host_template.sg_tablesize’)
rts51x_scsi.c:2125:24: error: initializer element is not computable at load time
rts51x_scsi.c:2125:24: note: (near initialization for ‘rts51x_host_template.sg_tablesize’)
rts51x_scsi.c:2126:10: error: ‘struct scsi_host_template’ has no member named ‘slave_configure’; did you mean ‘sdev_configure’?
 2126 |         .slave_configure = slave_configure,
      |          ^~~~~~~~~~~~~~~
      |          sdev_configure
rts51x_scsi.c:2126:28: error: initialization of ‘short unsigned int’ from ‘int (*)(struct scsi_device *)’ makes integer from pointer without a cast [-Wint-conversion]
 2126 |         .slave_configure = slave_configure,
      |                            ^~~~~~~~~~~~~~~
rts51x_scsi.c:2126:28: note: (near initialization for ‘rts51x_host_template.sg_prot_tablesize’)
rts51x_scsi.c:2126:28: error: initializer element is not computable at load time
rts51x_scsi.c:2126:28: note: (near initialization for ‘rts51x_host_template.sg_prot_tablesize’)
make[4]: *** [/usr/src/kernels/6.14.2-300.fc42.x86_64/scripts/Makefile.build:207: rts51x_scsi.o] Error 1
make[3]: *** [/usr/src/kernels/6.14.2-300.fc42.x86_64/Makefile:2034: .] Error 2
make[2]: *** [/usr/src/kernels/6.14.2-300.fc42.x86_64/Makefile:263: __sub-make] Error 2
make[2]: Leaving directory '/tmp/rts5139'
make[1]: *** [Makefile:263: __sub-make] Error 2
make[1]: Leaving directory '/usr/src/kernels/6.14.2-300.fc42.x86_64'
make: *** [Makefile:50: all] Error 2
                                                                                                                                            
rts5139 git:master  
❯ 
                                                                                                                                            
rts5139 git:master  
❯ uname -r
6.14.2-300.fc42.x86_64
                                                                                                                                            
rts5139 git:master  
❯ lsusb | grep "Card"
Bus 001 Device 004: ID 0bda:0129 Realtek Semiconductor Corp. RTS5129 Card Reader Controller
                                                                                                                                            
rts5139 git:master  
❯ 
```

but it compiles successfully on the same system and 6.13 kernel

--- a/rts51x_scsi.c
+++ b/rts51x_scsi.c
@@ -1923,7 +1923,7 @@ int rts51x_scsi_handler(struct scsi_cmnd *srb, struct rts51x_chip *chip)
  * Host functions
  ***********************************************************************/
 
-int slave_alloc(struct scsi_device *sdev)
+int sdev_init(struct scsi_device *sdev)
 {
 	/*
 	 * Set the INQUIRY transfer length to 36.  We don't use any of
@@ -1934,7 +1934,7 @@ int slave_alloc(struct scsi_device *sdev)
 	return 0;
 }
 
-int slave_configure(struct scsi_device *sdev)
+int sdev_configure(struct scsi_device *sdev, struct queue_limits *limits)
 {
 	/* Scatter-gather buffers (all but the last) must have a length
 	 * divisible by the bulk maxpacket size.  Otherwise a data packet
@@ -2122,8 +2122,8 @@ struct scsi_host_template rts51x_host_template = {
 	/* unknown initiator id */
 	.this_id = -1,
 
-	.slave_alloc = slave_alloc,
-	.slave_configure = slave_configure,
+	.sdev_init = sdev_init,
+	.sdev_configure = sdev_configure,
 
 	/* lots of sg segments can be handled */
 	.sg_tablesize = SG_ALL,
@@ -2171,8 +2171,8 @@ struct scsi_host_template rts51x_host_template = {
 	/* unknown initiator id */
 	.this_id = -1,
 
-	.slave_alloc = slave_alloc,
-	.slave_configure = slave_configure,
+	.sdev_init = sdev_init,
+	.sdev_configure = sdev_configure,
 
 	/* lots of sg segments can be handled */
 	.sg_tablesize = SG_ALL,

--- a/rts51x_scsi.c
+++ b/rts51x_scsi.c
@@ -1923,7 +1923,12 @@ int rts51x_scsi_handler(struct scsi_cmnd *srb, struct rts51x_chip *chip)
  * Host functions
  ***********************************************************************/
 
+/* slave_alloc() was renamed into sdev_init() in kernel 6.14 */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0))
 int sdev_init(struct scsi_device *sdev)
+#else
+int slave_alloc(struct scsi_device *sdev)
+#endif
 {
 	/*
 	 * Set the INQUIRY transfer length to 36.  We don't use any of
@@ -1934,7 +1939,12 @@ int sdev_init(struct scsi_device *sdev)
 	return 0;
 }
 
+/* slave_configure() was removed from kernel 6.14 */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0))
 int sdev_configure(struct scsi_device *sdev, struct queue_limits *limits)
+#else
+int slave_configure(struct scsi_device *sdev)
+#endif
 {
 	/* Scatter-gather buffers (all but the last) must have a length
 	 * divisible by the bulk maxpacket size.  Otherwise a data packet
@@ -2122,8 +2132,14 @@ struct scsi_host_template rts51x_host_template = {
 	/* unknown initiator id */
 	.this_id = -1,
 
+	/* slave_alloc() was renamed & slave_configure() was removed in kernel 6.14 */
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0))
 	.sdev_init = sdev_init,
 	.sdev_configure = sdev_configure,
+#else
+	.slave_alloc = slave_alloc,
+	.slave_configure = slave_configure,
+#endif
 
 	/* lots of sg segments can be handled */
 	.sg_tablesize = SG_ALL,
@@ -2171,8 +2187,8 @@ struct scsi_host_template rts51x_host_template = {
 	/* unknown initiator id */
 	.this_id = -1,
 
-	.sdev_init = sdev_init,
-	.sdev_configure = sdev_configure,
+	.slave_alloc = slave_alloc,
+	.slave_configure = slave_configure,
 
 	/* lots of sg segments can be handled */
 	.sg_tablesize = SG_ALL,

--- a/rts51x_scsi.h
+++ b/rts51x_scsi.h
@@ -32,6 +32,7 @@
 #include <linux/blkdev.h>
 #include <linux/completion.h>
 #include <linux/mutex.h>
+#include <linux/version.h>
 #include <scsi/scsi_host.h>
 
 #include "rts51x_chip.h"
@@ -145,8 +146,14 @@ struct Scsi_Host;
 struct scsi_device;
 struct scsi_cmnd;
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 14, 0))
 int sdev_init(struct scsi_device *sdev);
 int sdev_configure(struct scsi_device *sdev, struct queue_limits *limits);
+#else
+int slave_alloc(struct scsi_device *sdev);
+int slave_configure(struct scsi_device *sdev);
+#endif
+
 int queuecommand(struct Scsi_Host *, struct scsi_cmnd *);
 int command_abort(struct scsi_cmnd *srb);
 int bus_reset(struct scsi_cmnd *srb);

--- a/rts51x_scsi.h
+++ b/rts51x_scsi.h
@@ -145,8 +145,8 @@ struct Scsi_Host;
 struct scsi_device;
 struct scsi_cmnd;
 
-int slave_alloc(struct scsi_device *sdev);
-int slave_configure(struct scsi_device *sdev);
+int sdev_init(struct scsi_device *sdev);
+int sdev_configure(struct scsi_device *sdev, struct queue_limits *limits);
 int queuecommand(struct Scsi_Host *, struct scsi_cmnd *);
 int command_abort(struct scsi_cmnd *srb);
 int bus_reset(struct scsi_cmnd *srb);

